test/controllers/products_controller_test.rb

```diff
@@ -31,4 +41,86 @@
       expect(flash[:error]).must_equal "Unknown product"
     end
   end
+
+  describe "new" do
+    describe "logged in merchant" do
+      it "succeeds" do
+        perform_login
+        get new_product_path
+
+        must_respond_with :success
+      end
+    end
+    describe "not logged in" do
+      it "redirects user" do
+        get new_product_path
+
+        must_respond_with :redirect
+        expect(flash[:error]).must_equal "You must be logged in to add a new product"
+      end
+    end
+  end
+
+  describe "create" do
+    describe "logged in user" do
+      it "creates a product with valid data" do
+        merchant = perform_login
+        new_product = {product: {name: "Something amazing", merchant_id: merchant.id, price: 1000}}
+        expect {
+          post products_path, params: new_product
+        }.must_change "Product.count", 1
+
+        new_product_id = Product.find_by(name: "Something amazing").id
+
+        must_respond_with :redirect
+        must_redirect_to product_path(new_product_id)
+      end
+    end
+
+    it "renders bad_request and does not update the DB for bogus data" do
+      merchant = perform_login
+      bad_prod_name = {product: {name: nil, merchant_id: merchant.id, price: 1000}}
+      bad_merch_id = {product: {name: "Awesome product", merchant_id: nil, price: 1000}}
+      bad_price = {product: {name: "Awesomer product", merchant_id: merchant.id, price: 0}}
+
+      expect {
+        post products_path, params: bad_prod_name
+      }.wont_change "Product.count"
+
+      must_respond_with :bad_request
+      expect(flash[:error]).must_include "Could not add new product"
+      expect(flash[:error]).must_include "name"
+      expect(flash[:error]).must_include "can't be blank"
+
+      expect {
+        post products_path, params: bad_merch_id
+      }.wont_change "Product.count"
+
+      must_respond_with :bad_request
+      expect(flash[:error]).must_include "Could not add new product"
+      expect(flash[:error]).must_include "merchant"
+      expect(flash[:error]).must_include "must exist"
+
+      expect {
+        post products_path, params: bad_price
+      }.wont_change "Product.count"
+
+      must_respond_with :bad_request
+      expect(flash[:error]).must_include "Could not add new product"
+      expect(flash[:error]).must_include "must be greater than 0"
```

Comment:
This is a very robust test! Chris said that for validations we only need to test one of them. I'm certainly not opposed to testing them all! Just wanted to mention it in case it saves you some work in the future. :-)